File: src/sun-math.mjs
// src/sun-math.mjs — pure Mathematik (kein DOM/Three)
export const DEG = Math.PI / 180;
export const RAD = 180 / Math.PI;

// Sonnenstand (kompakte NOAA/SunCalc-Variante).
// Azimut von Norden im Uhrzeigersinn, Höhe in Radiant.
export function sunPosition(date, lat, lng) {
  const dayMs = 86400000, J1970 = 2440588, J2000 = 2451545;
  const e = 23.4397 * DEG;
  const d = date.valueOf() / dayMs - 0.5 + J1970 - J2000;
  const M = (357.5291 + 0.98560028 * d) * DEG;
  const C = (1.9148 * Math.sin(M) + 0.02 * Math.sin(2 * M) + 0.0003 * Math.sin(3 * M)) * DEG;
  const P = 102.9372 * DEG;
  const L = M + C + P + Math.PI;
  const dec = Math.asin(Math.sin(e) * Math.sin(L));
  const ra  = Math.atan2(Math.cos(e) * Math.sin(L), Math.cos(L));
  const lw  = -lng * DEG;
  const phi = lat * DEG;
  const theta = (280.16 + 360.9856235 * d) * DEG - lw;
  const H = theta - ra;
  const azS = Math.atan2(Math.sin(H), Math.cos(H) * Math.sin(phi) - Math.tan(dec) * Math.cos(phi));
  let azN = azS + Math.PI;
  azN = (azN % (2 * Math.PI) + 2 * Math.PI) % (2 * Math.PI);
  const alt = Math.asin(Math.sin(phi) * Math.sin(dec) + Math.cos(phi) * Math.cos(dec) * Math.cos(H));
  return { azimuth: azN, altitude: alt };
}

// Einheitsvektor zur Sonne. +X Ost, +Y oben, −Z Nord.
export function sunVector(azN, alt) {
  const ch = Math.cos(alt);
  return {
    x: Math.sin(azN) * ch,
    y: Math.sin(alt),
    z: -Math.cos(azN) * ch,
  };
}

// 2D-Rotation in der x/z-Ebene um Winkel a (rad), positiv im UZS um +Y.
function rotY(p, a) {
  const c = Math.cos(a), s = Math.sin(a);
  return { x: p.x * c + p.z * s, y: p.y, z: -p.x * s + p.z * c };
}
// Rotation um X-Achse um Winkel a (rad).
function rotX(p, a) {
  const c = Math.cos(a), s = Math.sin(a);
  return { x: p.x, y: p.y * c - p.z * s, z: p.y * s + p.z * c };
}

// Vier Welt-Eckpunkte des Dach-Rechtecks nach Yaw → Neigungsrichtung → Neigung.
export function rectCornersWorld({ L, B, yawDeg, tiltDeg, tiltDirDeg, height, eye, front }) {
  const yaw = yawDeg * DEG, dir = tiltDirDeg * DEG, tip = tiltDeg * DEG;
  const hx = L / 2, hz = B / 2;
  const local = [
    { x:  hx, y: 0, z:  hz },
    { x:  hx, y: 0, z: -hz },
    { x: -hx, y: 0, z: -hz },
    { x: -hx, y: 0, z:  hz },
  ];
  const topY = -eye + height;
  return local.map((p) => {
    // Reihenfolge: Yaw → Neigungsrichtung → Neigungswinkel
    let q = rotY(p, yaw);
    q = rotY(q, dir);
    q = rotX(q, tip);
    return { x: q.x, y: q.y + topY, z: q.z - front };
  });
}

// Projektion eines Punktes entlang des Sonnenvektors auf die Bodenebene y = yGround.
export function projectToGround(p, sunVec, yGround) {
  const t = (p.y - yGround) / sunVec.y;
  return { x: p.x - sunVec.x * t, z: p.z - sunVec.z * t };
}

// Fläche eines Polygons in der x/z-Ebene (Shoelace, immer ≥ 0).
export function polygonArea(pts) {
  let a = 0;
  for (let i = 0; i < pts.length; i++) {
    const p = pts[i], q = pts[(i + 1) % pts.length];
    a += p.x * q.z - q.x * p.z;
  }
  return Math.abs(a) / 2;
}

// Schattenlänge des Masts: Höhe / tan(Sonnenhöhe).
export function mastShadowLength(height, altitudeRad) {
  if (altitudeRad <= 0) return Infinity;
  return height / Math.tan(altitudeRad);
}

// Numerische Schattenkennwerte des Dachs (Fläche + Mastlänge).
export function shadowMetrics({ L, B, yawDeg, tiltDeg, tiltDirDeg, height, eye, front, azimuthRad, altitudeRad }) {
  if (altitudeRad <= 0) return { areaM2: 0, lengthM: Infinity, isNight: true };
  const sv = sunVector(azimuthRad, altitudeRad);
  const yGround = -eye;
  const corners = rectCornersWorld({ L, B, yawDeg, tiltDeg, tiltDirDeg, height, eye, front });
  const ground = corners.map((c) => projectToGround(c, sv, yGround));
  return {
    areaM2: polygonArea(ground),
    lengthM: mastShadowLength(height, altitudeRad),
    isNight: false,
  };
}

File: ios/SunMathKit/Sources/SunMathKit/SunMath.swift
import Foundation
import simd

public enum SunMath {
    public static let deg = Double.pi / 180

    public static func position(date: Date, lat: Double, lng: Double) -> (azimuth: Double, altitude: Double) {
        let dayMs = 86_400_000.0, j1970 = 2_440_588.0, j2000 = 2_451_545.0
        let e = 23.4397 * deg
        let d = date.timeIntervalSince1970 * 1000.0 / dayMs - 0.5 + j1970 - j2000
        let m = (357.5291 + 0.98560028 * d) * deg
        let c = (1.9148 * sin(m) + 0.02 * sin(2 * m) + 0.0003 * sin(3 * m)) * deg
        let p = 102.9372 * deg
        let l = m + c + p + .pi
        let dec = asin(sin(e) * sin(l))
        let ra = atan2(cos(e) * sin(l), cos(l))
        let lw = -lng * deg
        let phi = lat * deg
        let theta = (280.16 + 360.9856235 * d) * deg - lw
        let h = theta - ra
        let azS = atan2(sin(h), cos(h) * sin(phi) - tan(dec) * cos(phi))
        var azN = azS + .pi
        azN = (azN.truncatingRemainder(dividingBy: 2 * .pi) + 2 * .pi).truncatingRemainder(dividingBy: 2 * .pi)
        let alt = asin(sin(phi) * sin(dec) + cos(phi) * cos(dec) * cos(h))
        return (azN, alt)
    }

    public static func vector(azimuth: Double, altitude: Double) -> SIMD3<Double> {
        let ch = cos(altitude)
        return SIMD3(sin(azimuth) * ch, sin(altitude), -cos(azimuth) * ch)
    }

    private static func rotY(_ p: SIMD3<Double>, _ a: Double) -> SIMD3<Double> {
        let c = cos(a), s = sin(a)
        return SIMD3(p.x * c + p.z * s, p.y, -p.x * s + p.z * c)
    }

    private static func rotX(_ p: SIMD3<Double>, _ a: Double) -> SIMD3<Double> {
        let c = cos(a), s = sin(a)
        return SIMD3(p.x, p.y * c - p.z * s, p.y * s + p.z * c)
    }

    public static func rectCornersWorld(L: Double, B: Double, yawDeg: Double, tiltDeg: Double,
                                        tiltDirDeg: Double, height: Double, eye: Double, front: Double) -> [SIMD3<Double>] {
        let yaw = yawDeg * deg, dir = tiltDirDeg * deg, tip = tiltDeg * deg
        let hx = L / 2, hz = B / 2
        let local = [SIMD3(hx,0,hz), SIMD3(hx,0,-hz), SIMD3(-hx,0,-hz), SIMD3(-hx,0,hz)]
        let topY = -eye + height
        return local.map { p in
            var q = rotY(p, yaw); q = rotY(q, dir); q = rotX(q, tip)
            return SIMD3(q.x, q.y + topY, q.z - front)
        }
    }

    /// World-space rim points of a round canopy (radius `radius`) at the mast
    /// top, after tilt-direction and tilt. Yaw is irrelevant for a disc.
    public static func roundRimWorld(radius: Double, tiltDeg: Double, tiltDirDeg: Double,
                                     height: Double, eye: Double, front: Double, segments: Int) -> [SIMD3<Double>] {
        let dir = tiltDirDeg * deg, tip = tiltDeg * deg
        let topY = -eye + height
        let n = max(3, segments)
        return (0..<n).map { i in
            let a = 2 * Double.pi * Double(i) / Double(n)
            let local = SIMD3(radius * cos(a), 0, radius * sin(a))
            var q = rotY(local, dir); q = rotX(q, tip)
            return SIMD3(q.x, q.y + topY, q.z - front)
        }
    }

    /// Projected ground outline (x,z) of the canopy's shadow for the given sun.
    /// Returns an empty array at night (altitude ≤ 0). For round canopies the
    /// radius is derived from `area`; `segments` controls the round resolution.
    public static func shadowGroundOutline(isRound: Bool, L: Double, B: Double, area: Double,
                                           yawDeg: Double, tiltDeg: Double, tiltDirDeg: Double,
                                           height: Double, eye: Double, front: Double,
                                           azimuth: Double, altitude: Double, segments: Int) -> [SIMD2<Double>] {
        if altitude <= 0 { return [] }
        let sv = vector(azimuth: azimuth, altitude: altitude)
        let worldPts: [SIMD3<Double>]
        if isRound {
            let r = (area / Double.pi).squareRoot()
            worldPts = roundRimWorld(radius: r, tiltDeg: tiltDeg, tiltDirDeg: tiltDirDeg,
                                     height: height, eye: eye, front: front, segments: segments)
        } else {
            worldPts = rectCornersWorld(L: L, B: B, yawDeg: yawDeg, tiltDeg: tiltDeg,
                                        tiltDirDeg: tiltDirDeg, height: height, eye: eye, front: front)
        }
        return worldPts.map { projectToGround($0, sun: sv, yGround: -eye) }
    }

    public static func projectToGround(_ p: SIMD3<Double>, sun: SIMD3<Double>, yGround: Double) -> SIMD2<Double> {
        let t = (p.y - yGround) / sun.y
        return SIMD2(p.x - sun.x * t, p.z - sun.z * t)
    }

    public static func polygonArea(_ pts: [SIMD2<Double>]) -> Double {
        var a = 0.0
        for i in 0..<pts.count {
            let p = pts[i], q = pts[(i + 1) % pts.count]
            a += p.x * q.y - q.x * p.y   // SIMD2: .x=x, .y=z
        }
        return abs(a) / 2
    }

    public static func mastShadowLength(height: Double, altitude: Double) -> Double {
        altitude <= 0 ? .infinity : height / tan(altitude)
    }

    public static func shadowMetrics(L: Double, B: Double, yawDeg: Double, tiltDeg: Double, tiltDirDeg: Double,
                                     height: Double, eye: Double, front: Double,
                                     azimuth: Double, altitude: Double) -> (areaM2: Double, lengthM: Double, isNight: Bool) {
        if altitude <= 0 { return (0, .infinity, true) }
        let sv = vector(azimuth: azimuth, altitude: altitude)
        let corners = rectCornersWorld(L: L, B: B, yawDeg: yawDeg, tiltDeg: tiltDeg,
                                       tiltDirDeg: tiltDirDeg, height: height, eye: eye, front: front)
        let ground = corners.map { projectToGround($0, sun: sv, yGround: -eye) }
        return (polygonArea(ground), mastShadowLength(height: height, altitude: altitude), false)
    }
}
